### ISSUE TITLE
Content-Length: 0 in POST Request with empty body

### ISSUE THREAD

**Issue by NghiaTranUIT** [apple/swift-nio#1915]
### Description
First of all, thanks to Apple Team who develops and maintains such a powerful network library 👍 

I've encountered the problem that when I make a POST Request (Empty Body) to `http://httpbin.org:80`, NIOHTTP1 `HTTPClientHandler` automatically adds the header `Content-Length: 0`.

- Why this header is added to the Request, even though the body is empty?
- Are there any way to opt-out of this behavior? I tried to set a breakpoint on `HTTPRequestEncoder`, but I could not find anywhere that this header is added.

---------------

Regarding the context, I'm working on the mitmproxy app (Proxyman). When the client sends a Post request (Empty body) with no Content-Length header to Proxyman, Proxyman will forwards the data package to the server. However, NIO automatically adds Content-Length: 0 headers, which accidentally modify the original request.

---------------

Thanks for your time 🙇 

### Expected behavior
Should not add `Content-Length: 0` in the POST Request Header or there is a way to **opt-out** of this behavior.

### Actual behavior
`Content-Length: 0` header is automatically added to a Post request, even though the body is empty

### Steps to reproduce

1. Run the example code. It will make an HTTP Post Request (Empty Body) to `http://httpbin.org:80`
2. Observe the Console log, HTTPBin returns the `Content-Length: 0` => It indicates that the out-going Request has "Content-Length: 0" in the header

### If possible, minimal yet complete reproducer code (or URL to code)

```swift
import NIO
import NIOHTTP1

private final class HTTPEchoHandler: ChannelInboundHandler {
    public typealias InboundIn = HTTPClientResponsePart
    public typealias OutboundOut = HTTPClientRequestPart
    
    public func channelActive(context: ChannelHandlerContext) {
        print("Client connected to \(context.remoteAddress!)")

        // Post with empty body
        let headers = HTTPHeaders([("Host", "httpbin.org")])
        let requestHead = HTTPRequestHead(version: .http1_1,
                                          method: .POST,
                                          uri: "/post",
                                          headers: headers)
        
        context.write(self.wrapOutboundOut(.head(requestHead)), promise: nil)
        context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
    }

    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
        let clientResponse = self.unwrapInboundIn(data)
        switch clientResponse {
        case .head(let responseHead):
            print("Received head: \(responseHead)")
        case .body(let byteBuffer):
            let string = String(buffer: byteBuffer)
            print("Received: '\(string)' back from the server.")
        case .end:
            print("Closing channel.")
            context.close(promise: nil)
        }
    }

    public func errorCaught(context: ChannelHandlerContext, error: Error) {
        context.close(promise: nil)
    }
}

let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
let bootstrap = ClientBootstrap(group: group)
    // Enable SO_REUSEADDR.
    .channelOption(ChannelOptions.socketOption(.so_reuseaddr), value: 1)
    .channelInitializer { channel in
        channel.pipeline.addHTTPClientHandlers().flatMap {
            channel.pipeline.addHandler(HTTPEchoHandler())
        }
    }
defer {
    try! group.syncShutdownGracefully()
}

let defaultHost = "httpbin.org"
let defaultPort: Int = 80
let channel = try { () -> Channel in
    return try bootstrap.connect(host: defaultHost, port: defaultPort).wait()
}()

// Will be closed after we echo-ed back to the server.
try channel.closeFuture.wait()

print("Client closed")
```

**Output**
```
Client connected to [IPv4]52.201.75.114/52.201.75.114:80
Received head: HTTPResponseHead { version: HTTP/1.1, status: ok, headers: [("Date", "Fri, 23 Jul 2021 10:12:35 GMT"), ("Content-Type", "application/json"), ("Content-Length", "290"), ("Connection", "keep-alive"), ("Server", "gunicorn/19.9.0"), ("Access-Control-Allow-Origin", "*"), ("Access-Control-Allow-Credentials", "true")] }
Received: '{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "0", 
    "Host": "httpbin.org", 
    "X-Amzn-Trace-Id": "Root=1-60fa9613-0f26ed076a55b07b042bb9a2"
  }, 
  "json": null, 
  "origin": "27.75.135.126", 
  "url": "http://httpbin.org/post"
}
```

### SwiftNIO version/commit hash

SwiftNIO 2.30.0

### Swift & OS version (output of `swift --version && uname -a`)
Apple Swift version 5.4.2 (swiftlang-1205.0.28.2 clang-1205.0.19.57)
Target: arm64-apple-darwin20.6.0
Darwin Nghias-Mac-mini.local 20.6.0 Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:27 PDT 2021; root:xnu-7195.141.2~5/RELEASE_ARM64_T8101 arm64

**Comment by Lukasa** [comment]
Thanks for the report!

The HTTP/1.1 framing headers are modified by the method `correctlyFrameTransportHeaders`: https://github.com/apple/swift-nio/blob/906d5c554bd074a2cfc1b1c1d9f92dfa0505ceda/Sources/NIOHTTP1/HTTPEncoder.swift#L95-L129

However, you rightly observe that this will never add a content-length header field. So how do we appear to have one?

Ultimately this isn't to do with NIO, but instead to do with httpbin. httpbin is a Python application running behind a WSGI server (it was running behind gunicorn last I checked, but it could be running behind something else now). WSGI does not support chunked transfer encoding, so in general WSGI servers will buffer up a chunked body into a single big blob and remove the chunked encoding header before passing it on. It seems that the WSGI server is doing this, and so an "empty" POST request with a chunked transfer encoding header is manifesting as a "content-length: 0" request.

Wireshark reveals this to be true. According to Wireshark, this is what your NIO request actually looks like when it was sent on the network:

```
POST /post HTTP/1.1
Host: httpbin.org
transfer-encoding: chunked

0

```

This is a chunked POST of zero size. However, httpbin sends this:

```
HTTP/1.1 200 OK
Date: Fri, 23 Jul 2021 10:48:33 GMT
Content-Type: application/json
Content-Length: 290
Connection: keep-alive
Server: gunicorn/19.9.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true

{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "0", 
    "Host": "httpbin.org", 
    "X-Amzn-Trace-Id": "Root=1-60fa9e81-22f20b65333842f3351b7179"
  }, 
  "json": null, 
  "origin": "195.99.45.158", 
  "url": "http://httpbin.org/post"
}
```

As you can see, this includes a few headers we didn't send (content-length and x-amzn-trace-id) which were added by the remote infrastructure itself.

The TL;DR here is: NIO never adds a content-length header. It _does_, however, add a transfer-encoding: chunked header field.

**Comment by NghiaTranUIT** [comment]
Thanks for your detailed answer @Lukasa 👍 

It looks like look like by adding the Transfer-Encoding header, the remote infrastructure will add `Content-Length: 0`.

Is there any way to omit to add the Transfer-encoding? I tried to add "Transfer-Encoding: identity", but the result is the same.

I would like to achieve the same output as the response of the following curl. No Content-Length.

```bash
$ curl -v --http1.1 --request POST \
  --url https://httpbin.org/post
```
Output:
```bash
➜  ~ 
> POST /post HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Sat, 24 Jul 2021 09:21:27 GMT
< Content-Type: application/json
< Content-Length: 319
< Connection: keep-alive
< Server: gunicorn/19.9.0
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
<
{
  "args": {},
  "data": "",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.org",
    "User-Agent": "curl/7.64.1",
    "X-Amzn-Trace-Id": "Root=1-60fbdb97-7bba8021733738de177526db"
  },
  "json": null,
  "origin": "27.75.135.126",
  "url": "https://httpbin.org/post"
}
* Connection #0 to host httpbin.org left intact
* Closing connection 0
* ```

**Comment by Lukasa** [comment]
NIO currently does not have a mode where it will just pass through the transport headers as they are. It certainly could, and we are open to adding that API surface to the `HTTPRequestEncoder` and tolerating POST request bodies without payloads.

**Comment by NghiaTranUIT** [comment]
Thanks @Lukasa It's good to know that it could be fixed. I will fork the project and try implementing my change 👍 

Feel free to close the ticket 🙇
